pkg/driver/kubernetes_driver.go

```diff
@@ -0,0 +1,366 @@
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/deislabs/cnab-go/driver"
+	batchv1 "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	batchclientv1 "k8s.io/client-go/kubernetes/typed/batch/v1"
+	coreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	k8sContainerName    = "invocation"
+	k8sFileSecretVolume = "files"
+)
+
+// KubernetesDriver is capable of running an invocation image in a Kubernetes cluster.
+type KubernetesDriver struct {
+	Namespace             string
+	ServiceAccountName    string
+	LimitCPU              resource.Quantity
+	LimitMemory           resource.Quantity
+	ActiveDeadlineSeconds int64
+	BackoffLimit          int32
+	SkipCleanup           bool
+	skipJobStatusCheck    bool
+	jobs                  batchclientv1.JobInterface
+	secrets               coreclientv1.SecretInterface
+	pods                  coreclientv1.PodInterface
+	deletionPolicy        metav1.DeletionPropagation
+	requiredCompletions   int32
+}
+
+// NewKubernetesDriver initializes a Kubernetes driver.
+func NewKubernetesDriver(namespace, serviceAccount string, conf *rest.Config) (*KubernetesDriver, error) {
+	driver := &KubernetesDriver{
+		Namespace:          namespace,
+		ServiceAccountName: serviceAccount,
+	}
+	driver.setDefaults()
+	err := driver.setClient(conf)
+	if err != nil {
+		return driver, err
+	}
+
+	return driver, nil
+}
+
+// Config returns the Kubernetes driver configuration options
+func (k *KubernetesDriver) Config() map[string]string {
+	return map[string]string{
+		"KUBE_NAMESPACE":  "Increase verbosity. true, false are supported values",
+		"SERVICE_ACCOUNT": "Always pull image, even if locally available (0|1)",
+		"KUBE_CONFIG":     "Absolute path to the kubeconfig file",
+		"MASTER_URL":      "Kubernetes master endpoint",
+	}
+}
+
+// SetConfig sets Kubernetes driver configuration
+func (k *KubernetesDriver) SetConfig(settings map[string]string) {
+	k.setDefaults()
+	k.Namespace = settings["KUBE_NAMESPACE"]
+	k.ServiceAccountName = settings["SERVICE_ACCOUNT"]
+
+	var kubeconfig string
+	if kpath := settings["KUBE_CONFIG"]; kpath != "" {
+		kubeconfig = kpath
+	} else if home := homeDir(); home != "" {
+		kubeconfig = filepath.Join(home, ".kube", "config")
+	}
+
+	conf, err := clientcmd.BuildConfigFromFlags(settings["MASTER_URL"], kubeconfig)
+	if err != nil {
+		panic(err)
+	}
+	err = k.setClient(conf)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (k *KubernetesDriver) setDefaults() {
+	k.SkipCleanup = false
+	k.BackoffLimit = 0
+	k.ActiveDeadlineSeconds = 300
+	k.requiredCompletions = 1
+	k.deletionPolicy = metav1.DeletePropagationBackground
+}
+
+func (k *KubernetesDriver) setClient(conf *rest.Config) error {
+	coreClient, err := coreclientv1.NewForConfig(conf)
+	if err != nil {
+		return err
```

Comment:
Why return an error to a caller which panics? "First failure data capture" would be better by panicking here. Alternatively, errors could be returned everywhere after logging a suitable message and then panic just before the driver loses control (unless the caller of the driver panics if an error is returned). The latter approach is idiomatic Go and lends itself to easier unit testing.

Reply:
There are actually two callers -- `NewKubernetesDriver` does return the error:

https://github.com/deislabs/duffle/blob/206a2bec45ddddb9dcf558581f27249a80db835f/pkg/driver/kubernetes.go#L45-L53

The `Configurable` interface's `SetConfig` method signature doesn't have an error return, so I figured panicking was the best thing to do for now. A better solution would probably be to have `SetConfig` return an error since any operation accepting user input is likely to be fallible anyway.

Reply:
Let's open an issue in cnab-go for this.